sdks/java/extensions/timeseries/src/main/java/org/apache/beam/sdk/extensions/timeseries/transforms/OrderOutput.java

```diff
@@ -0,0 +1,388 @@
+            Window.<KV<TimeSeriesData.TSKey, TimeSeriesData.TSAccum>>into(
+                    FixedWindows.of(options.downSampleDuration()))
+                // TODO: DirectRunner not showing results with exact late date match
+                //.withAllowedLateness(options.downSampleDuration().plus(options.downSampleDuration()))
+                .withAllowedLateness(Duration.standardDays(1))
+                .discardingFiredPanes());
+  }
+
+  /**
+   * When a new key is seen (state == null) for the first time, we will create a timer to fire in
+   * the next window boundary. If this is not the first time the key is seen we check the ttl to see
+   * if a new timer is required. In-between timers firing, we will add all new elements to a List.
+   * lets have 3 elements coming in at various time and then there is NULL for forth time slice
+   * [t1,t2,t3, NULL] t1 arrives and we set timer to fire at t1.plus(downsample duration + fixed
+   * offset).
+   *
+   * <p>Then we have state.set(t1). t2 arrives and we add t1 to t2 as the previous value and we
+   * output the value state.set(t2)
+   *
+   * <p>t3 arrives and we add t2 to t3 as the previous value and we output the value state.set(t3)
+   *
+   * <p>at time 4 we have no entry, here we use the last known state which is t3 and we change the
+   * time values.
+   *
+   * <p>NOTE: Need to see if this is really needed In case there is more than one element arriving
+   * before the timer fires, we will also hold a List of elements If the list of elements is > 0
+   * then we loop through the core logic until the list is exhausted
+   */
+  @Experimental
+  public static class GetPreviousData
+      extends DoFn<
+          KV<TimeSeriesData.TSKey, TimeSeriesData.TSAccum>,
+          KV<TimeSeriesData.TSKey, TimeSeriesData.TSAccum>> {
+
+    TSConfiguration options;
+
+    private GetPreviousData(TSConfiguration configuration) {
+      this.options = configuration;
+    }
+
+    // Setup our state objects
+
+    @StateId("lastKnownState")
+    private final StateSpec<ValueState<TimeSeriesData.TSAccum>> lastKnownState =
+        StateSpecs.value(ProtoCoder.of(TimeSeriesData.TSAccum.class));
+
+    @StateId("holdingList")
+    private final StateSpec<BagState<TimeSeriesData.TSAccum>> holdingList =
+        StateSpecs.bag(ProtoCoder.of(TimeSeriesData.TSAccum.class));
+
+    @TimerId("alarm")
+    private final TimerSpec alarm = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    /**
+     * This is the simple path... A new element is here so we add it to the list of elements and set
+     * a timer
+     *
+     * @param c
+     * @param holdingList
+     * @param timer
+     */
+    @ProcessElement
+    public void processElement(
+        ProcessContext c,
+        @StateId("holdingList") BagState<TimeSeriesData.TSAccum> holdingList,
+        @TimerId("alarm") Timer timer) {
+
+      Instant alarm = c.timestamp().plus(options.downSampleDuration());
+
+      timer.set(alarm);
```

Comment:
Do I understand it right that for all incoming elements we will set alarms?

Reply:
My understanding  is that each element would update the same timer object, but yes for each element into OrderOutput we will have a update of the timer. However the library only calls OrderedOutput after a FixedWindow+Combiner so the number of elements updating the alarm is != the raw elements flowing into the library.